lib/src/react_client/private_utils.dart

```diff
@@ -0,0 +1,231 @@
+
+/// Converts a list of variadic children arguments to children that should be passed to ReactJS.
+///
+/// Returns:
+///
+/// - `null` if there are no args
+/// - the single child if only one was specified
+/// - otherwise, the same list of args, will all top-level children validated
+dynamic convertArgsToChildren(List childrenArgs) {
+  if (childrenArgs.isEmpty) {
+    return null;
+  } else if (childrenArgs.length == 1) {
+    return childrenArgs.single;
+  } else {
+    markChildrenValidated(childrenArgs);
+    return childrenArgs;
+  }
+}
+
+/// Convert packed event handler into wrapper and pass it only the Dart [SyntheticEvent] object converted from the
+/// [events.SyntheticEvent] event.
+convertEventHandlers(Map args) {
+  args.forEach((propKey, value) {
+    var eventFactory = eventPropKeyToEventFactory[propKey];
+    if (eventFactory != null && value != null) {
+      // Apply allowInterop here so that the function we store in `originalEventHandlers`
+      // is the same one we'll retrieve from the JS props.
+      var reactDartConvertedEventHandler = allowInterop((events.SyntheticEvent e, [_, __]) {
+        value(eventFactory(e));
+      });
+
+      args[propKey] = reactDartConvertedEventHandler;
+      originalEventHandlers[reactDartConvertedEventHandler] = value;
+    }
+  });
+}
+
+void convertRefValue(Map args) {
+  var ref = args['ref'];
+  if (ref is Ref) {
+    args['ref'] = ref.jsRef;
+  }
+}
+
+void convertRefValue2(Map args, {bool convertCallbackRefValue = true}) {
+  var ref = args['ref'];
+
+  if (ref is Ref) {
+    args['ref'] = ref.jsRef;
+    // If the ref is a callback, pass ReactJS a function that will call it
+    // with the Dart Component instance, not the ReactComponent instance.
+    //
+    // Use CallbackRef<Null> to check arity, since parameters could be non-dynamic, and thus
+    // would fail the `is CallbackRef<dynamic>` check.
+    // See https://github.com/dart-lang/sdk/issues/34593 for more information on arity checks.
+  } else if (ref is CallbackRef<Null> && convertCallbackRefValue) {
+    args['ref'] = allowInterop((dynamic instance) {
+      // Call as dynamic to perform dynamic dispatch, since we can't cast to CallbackRef<dynamic>,
+      // and since calling with non-null values will fail at runtime due to the CallbackRef<Null> typing.
+      if (instance is ReactComponent && instance.dartComponent != null) return (ref as dynamic)(instance.dartComponent);
+      return (ref as dynamic)(instance);
+    });
+  }
+}
+
+/// Util used with `registerComponent2` to ensure no important lifecycle
+/// events are skipped. This includes `shouldComponentUpdate`,
+/// `componentDidUpdate`, and `render` because they utilize
+/// `_updatePropsAndStateWithJs`.
+///
+/// Returns the list of lifecycle events to skip, having removed the
+/// important ones. If an important lifecycle event was set for skipping, a
+/// warning is issued.
+List<String> filterSkipMethods(List<String> methods) {
```

Comment:
This seems like it should live in `component_registration.dart` (as private) since that's the only place it's used and it's tightly coupled to that usage, and it doesn't have more general value.